File: exercism/go/proverb/proverb.go
package proverb

import "fmt"

func Proverb(rhyme []string) []string {
	res := make([]string, len(rhyme))
	if len(rhyme) == 0 {
		return res
	}
	for i := range rhyme[1:] {
		res[i] = fmt.Sprintf("For want of a %s the %s was lost.", rhyme[i], rhyme[i+1])
	}
	res[len(res)-1] = fmt.Sprintf("And all for the want of a %s.", rhyme[0])
	return res
}

File: exercism/zig/proverb/proverb.zig
const std = @import("std");
const fmt = std.fmt;
const mem = std.mem;

pub fn recite(allocator: mem.Allocator, words: []const []const u8) (fmt.AllocPrintError || mem.Allocator.Error)![][]u8 {
    const len = words.len;
    if (len == 0) return &.{};
    const proverb = try allocator.alloc([]u8, len);
    var lines_allocated: usize = 0;
    errdefer {
        for (proverb[0..lines_allocated]) |line| allocator.free(line);
        allocator.free(proverb);
    }
    for (words[1..], 1..) |word, i| {
        proverb[i - 1] = try fmt.allocPrint(allocator, "For want of a {s} the {s} was lost.\n", .{ words[i - 1], word });
        lines_allocated += 1;
    }
    proverb[len - 1] = try fmt.allocPrint(allocator, "And all for the want of a {s}.\n", .{words[0]});
    return proverb;
}
